src/index.ts

```diff
@@ -94,46 +84,25 @@ export const createRouter = <
   ): void =>
     history.replace(matchToHistoryPath(matchers[routeName], first(args)));
 
-  const subscribe = (subscription: Subscription): (() => void) => {
-    subscriptions.add(subscription);
-
-    return () => {
-      subscriptions.delete(subscription);
-    };
-  };
-
-  const useLocation = (): Location =>
-    useSubscription(
-      React.useMemo(
-        () => ({ getCurrentValue: () => currentLocation, subscribe }),
-        [],
-      ),
-    );
-
   const useRoute = <RouteName extends keyof FiniteRoutes | keyof NestedRoutes>(
     routeNames: ReadonlyArray<RouteName>,
   ): RouteName extends string
     ? { name: RouteName; params: Simplify<RoutesParams[RouteName]> } | undefined
     : never => {
-    // JSON.{stringify,parse} is used to prevent some re-renders,
-    // as the params object instance is updated on each one
-    const route = useSubscription(
-      React.useMemo(() => {
-        const matchers = rankedMatchers.filter(({ name }) =>
-          routeNames.includes(name as RouteName),
-        );
-
-        return {
-          getCurrentValue: () => {
-            const route = match(currentLocation, matchers);
-            return route ? JSON.stringify(route) : route;
-          },
-          subscribe,
-        };
-      }, [JSON.stringify(routeNames)]),
-    );
-
-    return route ? JSON.parse(route) : route;
+    const location = useLocation();
+
+    const route = React.useMemo(() => {
+      const matchers = rankedMatchers.filter(({ name }) =>
+        routeNames.includes(name as RouteName),
+      );
+      return match(location, matchers);
+    }, [location, JSON.stringify(routeNames)]) as RouteName extends string
```

Comment:
Same issue here, it will trigger a re-render even if the matched route doesn't change. If the top level `useRoute` from your app listen to a wilcard route (`/projects/:projectId/*`), the `location` update will update the component even if the matched route is 100% identical